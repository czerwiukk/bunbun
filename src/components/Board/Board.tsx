import { faPlus, faTableColumns } from "@fortawesome/free-solid-svg-icons";
import Fa from "solid-fa";
import { Component, createResource, createSignal, For } from "solid-js";
import { changeTaskColumn } from "~/api";
import { BoardColumn } from "../BoardColumn/BoardColumn";
import {
  DragDropProvider,
  DragDropSensors,
  DragEvent,
} from "@thisbeyond/solid-dnd";
import { fetchBoardTasks } from "~/api/columns";
import { usePendingChanges, useColumns } from "~/hooks";
interface BoardProps {
  id: number;
}

export const Board: Component<BoardProps> = (props) => {
  const { boardColumns, createColumn, deleteColumn } = useColumns(
    () => props.id
  );

  const [boardTasks, { refetch: refetchTasks }] = createResource(
    () => props.id,
    fetchBoardTasks
  );

  const [isAddBtnMoved, moveAddBtn] = createSignal(false);

  const { getColumnTasksPredicate, addPendingChange, removePendingChange } =
    usePendingChanges;

  const onDragEnd = async ({ droppable, draggable }: DragEvent) => {
    if (!droppable || !draggable) return;
    const droppableId = +droppable.id;
    const draggableId = +draggable.id;
    addPendingChange(droppableId, draggableId);
    try {
      await changeTaskColumn(droppableId, draggableId);
    } finally {
      await refetchTasks();
      removePendingChange(droppableId);
    }
  };

  const createNewColumn = async () => {
    try {
      moveAddBtn(true);
      await createColumn();
    } finally {
      moveAddBtn(false);
    }
  };

  return (
    <div class="mt-8 flex flex-grow gap-4">
      <DragDropProvider onDragEnd={onDragEnd}>
        <DragDropSensors />
        <For each={boardColumns()}>
          {(column) => (
            <BoardColumn
              column={column}
              tasks={
                boardTasks()?.filter(getColumnTasksPredicate(column.id)) ?? []
              }
              onDeleteColumn={deleteColumn}
            />
          )}
        </For>
      </DragDropProvider>

      <div
        classList={{ "transition-transform translate-x-72": isAddBtnMoved() }}
      >
        <button
          class="btn-sm btn flex items-center gap-2 border-0 bg-transparent bg-amber-100 text-stone-900 hover:bg-amber-200 dark:bg-stone-800 dark:text-stone-100 dark:hover:bg-stone-700"
          onClick={createNewColumn}
        >
          <Fa icon={faTableColumns} />

          <Fa icon={faPlus} />
        </button>
      </div>
    </div>
  );
};
