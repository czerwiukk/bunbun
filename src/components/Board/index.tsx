import { faPlus, faTableColumns } from "@fortawesome/free-solid-svg-icons";
import Fa from "solid-fa";
import {
  Component,
  createEffect,
  createResource,
  createSignal,
  For,
} from "solid-js";
import { addColumn, deleteColumn, fetchBoardColumns } from "~/api";
import { BoardColumn } from "../BoardColumn";
import { DragDropProvider, DragDropSensors } from "@thisbeyond/solid-dnd";
import classNames from "classnames";
interface BoardProps {
  id: number;
}

export const Board: Component<BoardProps> = (props) => {
  const [boardColumns, { refetch: refetchColumns }] = createResource(
    () => props.id,
    fetchBoardColumns
  );
  const [where, setWhere] = createSignal("outside");
  const [isAddBtnMoved, moveAddBtn] = createSignal(false);

  createEffect(() => {
    console.log("board;", props.id);
  });

  const onDragEnd = ({ droppable }: any) => {
    if (droppable) {
      setWhere("inside");
    } else {
      setWhere("outside");
    }
  };

  const createNewColumn = async () => {
    try {
      moveAddBtn(true);
      await addColumn(props.id);
      await refetchColumns();
    } finally {
      moveAddBtn(false);
    }
  };
  const deleteExistingColumn = async (columnId: number) => {
    await deleteColumn(columnId);
    await refetchColumns();
  };

  return (
    <div class={classNames("flex gap-4 mt-8 flex-grow")}>
      <DragDropProvider onDragEnd={onDragEnd}>
        <DragDropSensors />
        <For each={boardColumns()}>
          {(column) => (
            <BoardColumn
              column={column}
              onDeleteColumn={deleteExistingColumn}
            />
          )}
        </For>
      </DragDropProvider>

      <div
        classList={{ "transition-transform translate-x-72": isAddBtnMoved() }}
      >
        <button
          class={classNames(
            "btn btn-sm bg-transparent bg-amber-50 hover:bg-amber-100  text-stone-900 border-0 flex items-center gap-2"
          )}
          onClick={createNewColumn}
        >
          <Fa icon={faTableColumns} />

          <Fa icon={faPlus} />
        </button>
      </div>
    </div>
  );
};
