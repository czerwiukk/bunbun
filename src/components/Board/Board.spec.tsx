import { render } from "solid-testing-library";
import { describe, expect, it, vi } from "vitest";
import { Board } from "./Board";

vi.mock("~/hooks", () => ({
  useColumns: () => ({}),
  usePendingChanges: {},
}));

vi.mock("~/api/columns", () => ({
  fetchBoardTasks: () => [],
}));

describe("Board", () => {
  it("matches snapshot", () => {
    expect(render(() => <Board id={1} />)).toMatchSnapshot();
  });
});
