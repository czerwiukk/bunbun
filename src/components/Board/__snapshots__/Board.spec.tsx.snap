// Vitest Snapshot v1

exports[`Board > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex gap-4 mt-8 flex-grow"
      >
        
        <div>
          <button
            class="btn btn-sm bg-transparent bg-amber-100 dark:bg-stone-800 hover:bg-amber-200 dark:hover:bg-stone-700 text-stone-900 dark:text-stone-100 border-0 flex items-center gap-2"
          >
            <svg
              aria-hidden="true"
              class="fa"
              style="height: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(256 256)"
                transform-origin="128 0"
              >
                <g
                  transform="translate(0,0) scale(1,1)"
                >
                  <path
                    d="M0 96C0 60.7 28.7 32 64 32H448c35.3 0 64 28.7 64 64V416c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V96zm64 64V416H224V160H64zm384 0H288V416H448V160z"
                    fill="currentColor"
                    transform="translate(-256 -256)"
                  />
                </g>
              </g>
            </svg>
            <svg
              aria-hidden="true"
              class="fa"
              style="height: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(224 256)"
                transform-origin="112 0"
              >
                <g
                  transform="translate(0,0) scale(1,1)"
                >
                  <path
                    d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
                    fill="currentColor"
                    transform="translate(-224 -256)"
                  />
                </g>
              </g>
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex gap-4 mt-8 flex-grow"
    >
      
      <div>
        <button
          class="btn btn-sm bg-transparent bg-amber-100 dark:bg-stone-800 hover:bg-amber-200 dark:hover:bg-stone-700 text-stone-900 dark:text-stone-100 border-0 flex items-center gap-2"
        >
          <svg
            aria-hidden="true"
            class="fa"
            style="height: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              transform="translate(256 256)"
              transform-origin="128 0"
            >
              <g
                transform="translate(0,0) scale(1,1)"
              >
                <path
                  d="M0 96C0 60.7 28.7 32 64 32H448c35.3 0 64 28.7 64 64V416c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V96zm64 64V416H224V160H64zm384 0H288V416H448V160z"
                  fill="currentColor"
                  transform="translate(-256 -256)"
                />
              </g>
            </g>
          </svg>
          <svg
            aria-hidden="true"
            class="fa"
            style="height: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              transform="translate(224 256)"
              transform-origin="112 0"
            >
              <g
                transform="translate(0,0) scale(1,1)"
              >
                <path
                  d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
                  fill="currentColor"
                  transform="translate(-224 -256)"
                />
              </g>
            </g>
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "unmount": [Function],
}
`;
